src/sqlfluff/core/parser/grammar/sequence.py

```diff
@@ -72,18 +111,46 @@ def match(self, segments, parse_context):
                 break
 
             while True:
-                # Consume non-code if appropriate
-                if self.allow_gaps:
-                    pre_nc, mid_seg, post_nc = trim_non_code_segments(
-                        unmatched_segments
+                # Is there anything left to match on?
+                if len(unmatched_segments) == 0:
+                    # There isn't, but we still have elements left to match.
+                    # Do only optional or meta elements remain?
+                    remaining_metas = _all_remaining_metas(
+                        self._elements[idx:], parse_context
                     )
-                else:
-                    pre_nc = ()
-                    mid_seg = unmatched_segments
-                    post_nc = ()
+                    if remaining_metas is not None:
+                        # We're safe. Claim them and return.
+                        meta_post_nc += remaining_metas
+                        early_break = True
```

Comment:
Is this flag actually used anywhere?

Reply:
Yeah, it's used just up above as a way of "double breaking", i.e. a way to break both out of the `while` loop and _also` the `for` loop.